L46-core-tls-credential-API.md

```diff
@@ -0,0 +1,390 @@
+  explicit TlsServerCredentialsOptions(
+      std::shared_ptr<CertificateProviderInterface> certificate_provider)
+      : TlsCredentialsOptions() {
+    set_certificate_provider(certificate_provider);
+  }
+
+
+  // Sets option to request the certificates from the client.
+  // The default is GRPC_SSL_DONT_REQUEST_CLIENT_CERTIFICATE.
+  void set_cert_request_type(
+      grpc_ssl_client_certificate_request_type cert_request_type);
+}
+
+// Contains configurable options on the client side.
+// Client side doesn't need to always use certificate provider. When the
+// certificate provider is not set, we will use the root certificates stored
+// in the system default locations, and assume client won't provide any
+// identity certificates(single side TLS).
+class TlsChannelCredentialsOptions final : public TlsCredentialsOptions {
+ public:
+  // Sets the decision of whether to do a crypto check on the server certs.
+  // The default is true.
+  void set_verify_server_certs(bool verify_server_certs);
+
+// Builds TLS Credentials given TLS options.
+std::shared_ptr<ChannelCredentials> TlsCredentials(
+    const TlsChannelCredentialsOptions& options);
+}
+
+std::shared_ptr<ServerCredentials> TlsServerCredentials(
+    const grpc::experimental::TlsServerCredentialsOptions& options);
+```
+
+Here is an example of how to use this API:
+```c++
+/* Create option and set basic security primitives. */
+TlsChannelCredentialsOptions options = TlsChannelCredentialsOptions();
+options.set_verify_server_cert(true);
+options.set_min_tls_version(TLS1_2);
+options.set_max_tls_version(TLS1_3);
+
+// The core credentials APIs are still C, so this is still a grpc_channel_credential
+/* Create TLS channel credentials. */
+std::shared_ptr<ChannelCredentials> creds = TlsCredentials(options);
+```
+
+### Credential Reloading
+Credential reloading will be implemented via a `CertificateProvider`. This provider is responsible for sourcing key material and supplying it to the internal stack via `SetKeyMaterials`.
+We provide a `StaticDataCertificateProvider` and a `FileWatcherCertificateProvider` that should cover many use cases.
+If these providers do not support the intricacies for a specific use case, a user can provide their own implementations of the `CertificateProviderInterface`.
+
+```c
+/* Opaque types. */
+// A struct that stores the credential data presented to the peer in handshake
+// to show local identity. The private_key and certificate_chain must be PEM
+// encoded and should always match.
+struct IdentityKeyCertPair {
+  std::string private_key;
+  std::string certificate_chain;
+};
+
+
+/** ------------------------------------ Provider ------------------------------------ */
+/** Interface for a class that handles the process to fetch credential data.
+*/
+class CertificateProviderInterface {
+ public:
+  virtual ~CertificateProviderInterface() = default;
+  // Provider implementations MUST provide a WatchStatusCallback that will be
```

Comment:
It would be really valuable if, in the paragraph here, we can distinguish between (a) the API contract for WatchStatusCallback, and (b) the internal implementation details of how this API will be called by the TLS stack. I think this comment should focus on (a), and should not mention (b).

Reply:
I reworked this, PTAL.

Reply:
Resolving for now, but we may need to revisit this whole interface, as discussed yesterday.